modules/core/test/com/haulmont/cuba/core/MessagesTest.java

```diff
@@ -111,6 +112,25 @@ public void testMultiInclude() {
         assertEquals(msg3, "overridden");
     }
 
+    @Test
+    public void testIncludeDefaultLoc() {
+        Messages messages = AppBeans.get(Messages.class);
+        TestMessageTools messageTools = (TestMessageTools) AppBeans.get(MessageTools.class);
+        try {
+            Locale localeRu = Locale.forLanguageTag("ru");
+            String msg = messages.getMessage("com.haulmont.cuba.core.mp_test", "includedMsg", localeRu);
+            assertEquals("Included Message", msg);
```

Comment:
You can add a check on the number of runs to search for a message:
```
    appender.getMessages().clear();
    Locale localeRu = Locale.forLanguageTag("ru");
    String msg = messages.getMessage("com.haulmont.cuba.core.mp_test", "includedMsg", localeRu);
    assertEquals("Included Message", msg);
    assertEquals( 10, getSearchMessagesCount());

    appender.getMessages().clear();
    userSession.setLocale(localeRu);
    messages.clearCache();
    msg = messages.getMessage("com.haulmont.cuba.core.mp_test", "includedMsg", localeRu);
    assertEquals("Included Message RU", msg);
    assertEquals(10, getSearchMessagesCount());
```
It will help you figure out why you need additional check in `AbstractMessages#searchIncludes()` method:
```
    Locale userLocale = getUserLocale();
    if (truncatedLocale == null
            && userLocale.equals(locale)) {
        message = searchMessage(includePath, key, locale, userLocale, passedPacks);
    } else {
        message = searchMessage(includePath, key, locale, truncatedLocale, passedPacks);
    }
```

Reply:
Thank you for your suggestions.

> if (truncatedLocale == null && userLocale.equals(locale)) {...

I think this suggestion is incorrect. `Messages` component allows to specify locale explicitly in all methods (through method overloadnig). And if it is specified the locale of current user should not be used and actually it is not used in the current implementation. At the moment `AbstractMessages#getUserLocale` method is used only when locale is not provided by the caller.

If we use `getUserLocale` in `searchIncludes` method we will get inconsistent behaviour. Suppose we have a project with supported locales RU and EN and default locale is RU. In this case russian messages are stored in default pack files, this schema is used in platform components and by CUBA studio. Suppose also that for some special screens we provide only russian translation so we use `Messages` methods with RU locale argument in that screens. If `@include` key is used in default (russian) pack of the screen `Messages.getMessage(class, key, RU)` method returns english message for users that have default English locale (becase of `userLocale.equals(locale)` check). In my opinion it is a strange and incorrect result.

Another example of problems with `getUserLocale` is login dialog. It allows user to select locale and then send it in `Credentials` object to `AuthenticationProvider` service. In `AuthenticationProvider` this locale is used to get error messages. If we extend `AuthenticationProvider` and use `@include` key in its message pack it can work incorrectly because locale in Credentials object and in user session can differ. I implemented this example in [demo project](https://github.com/daring2/cuba-issues-demo):

![image](https://user-images.githubusercontent.com/11490702/83496607-04d68600-a4c2-11ea-9035-98bf40ded4bd.png)

As I understand you suggestions are about performance (number of searches). But I am not sure about correctness of your fix. Maybe I missed something...?

Reply:
@daring2 Thank you, for your explanation and test project. I have looked at the case described above and I think that you are right. It will be correct to use the `messageTools.getDefaultLocale()` instead of `getUserLocale()`, as you described above.

Reply:
@GlebDurygin, thank you!